sanic/websocket.py

```diff
@@ -70,73 +799,94 @@ def connection_lost(self, exc):
 
     def data_received(self, data):
         if self.websocket is not None:
-            # pass the data to the websocket protocol
             self.websocket.data_received(data)
         else:
-            try:
-                super().data_received(data)
-            except HttpParserUpgrade:
-                # this is okay, it just indicates we've got an upgrade request
-                pass
+            # Pass it to HttpProtocol handler first
+            # That will (hopefully) upgrade it to a websocket.
+            super().data_received(data)
 
-    def write_response(self, response):
+    def eof_received(self) -> Optional[bool]:
         if self.websocket is not None:
-            # websocket requests do not write a response
-            self.transport.close()
+            return self.websocket.eof_received()
         else:
-            super().write_response(response)
+            return False
 
-    async def websocket_handshake(self, request, subprotocols=None):
-        # let the websockets package do the handshake with the client
-        headers = {}
+    def close(self):
+        # Called by HttpProtocol at the end of connection_task
+        # If we've upgraded to websocket, we do our own closure
+        if self.websocket is not None:
+            self.websocket.fail_connection(1001)
+        else:
+            super().close()
+
+    def close_if_idle(self):
+        # Called by Sanic Server when shutting down
+        # If we've upgraded to websocket, shut it down
+        if self.websocket is not None:
+            if self.websocket.connection.state in (CLOSING, CLOSED):
+                return True
+            else:
+                return self.websocket.fail_connection(1001)
+        else:
+            return super().close_if_idle()
 
+    async def websocket_handshake(self, request, subprotocols=Optional[Sequence[str]]):
```

Comment:
I see that you copy-pasted and modified `ServerConnection.accept` instead of wrapping it.

I thought it would be manageable to call it and convert its return value into a `SanicException` when the response code isn't 101. Does websockets make this approach difficult? Can I help?

Reply:
I used my own websocket_handshake instead of `ServerConnection.accept` for several reasons:
1) `websockets.accept()` handles errors by returning `self.reject()` that creates custom `websockets` `http.Response` objects, that are not quite compatible with Sanic's response handler. Sanic works better with `raise InvalidUsage()`-style exception throwing (and utilizing sanic's built-in error formatting, logging, and error Response generation).
2) `websockets.accept()` builds a `Headers()` structure that is not quite compatible with Sanic's response handler.
3) I wanted to write the `101 Switching Protocols` success-response back to the SanicProtocol's `send()` method directly, without having to go via the `websockets` `http.Response` object.

Having said all that, after thinking about it talking it through, I see that it probably wouldn't be too hard to refactor this method to use `websockets.accept()` instead of this custom logic.

Reply:
That makes sense to me; up to you to pick your favorite strategy.

Going through websockets' abstractions will create a little bit of overhead but opening handshakes aren't supposed to be the bottleneck of websockets (unless the user misses the persistent aspect of WebSocket connections). So perhaps the maintainability gains are worth the API overhead.